NPE when logging diagnostics due to bLangFuncNode.returnTypeNode.pos is null

**Description:**

The compiler gives NPE for the following code sample.
```bal
public function foo() returns "dulmina" {
}
```

</details>
<details>
  <summary><b>stacktrace</b>  (<i>click to expand</i>)</summary>

```
Caused by: java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.diagnostic.DiagnosticComparator.compare(DiagnosticComparator.java:44)
	at org.wso2.ballerinalang.compiler.diagnostic.DiagnosticComparator.compare(DiagnosticComparator.java:38)
	at java.base/java.util.TreeMap.compare(TreeMap.java:1292)
	at java.base/java.util.TreeMap.put(TreeMap.java:536)
	at java.base/java.util.TreeSet.add(TreeSet.java:255)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.addDiagnostic(BLangPackage.java:277)
	at org.wso2.ballerinalang.compiler.diagnostic.BLangDiagnosticLog.storeDiagnosticInModule(BLangDiagnosticLog.java:255)
	at org.wso2.ballerinalang.compiler.diagnostic.BLangDiagnosticLog.reportDiagnostic(BLangDiagnosticLog.java:222)
	at org.wso2.ballerinalang.compiler.diagnostic.BLangDiagnosticLog.error(BLangDiagnosticLog.java:99)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.checkForExportableType(CodeAnalyzer.java:2289)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExportableTypeRef(CodeAnalyzer.java:2221)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:474)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:385)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeTopLevelNodes(CodeAnalyzer.java:373)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:365)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:163)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyze(CodeAnalyzer.java:354)
	at io.ballerina.projects.internal.CompilerPhaseRunner.codeAnalyze(CompilerPhaseRunner.java:225)
	at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:120)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:374)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:323)
	at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:168)
	at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:160)
```
</details>

--- a/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java
+++ b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java
@@ -1272,6 +1272,7 @@ public class BLangNodeTransformer extends NodeTransformer<BLangNode> {
     public BLangNode transform(SingletonTypeDescriptorNode singletonTypeDescriptorNode) {
         BLangFiniteTypeNode bLangFiniteTypeNode = new BLangFiniteTypeNode();
         BLangLiteral simpleLiteral = createSimpleLiteral(singletonTypeDescriptorNode.simpleContExprNode());
+        bLangFiniteTypeNode.pos = simpleLiteral.pos;
         bLangFiniteTypeNode.valueSpace.add(simpleLiteral);
         return bLangFiniteTypeNode;
     }
@@ -2807,7 +2808,7 @@ public class BLangNodeTransformer extends NodeTransformer<BLangNode> {
     private BLangTupleVariableDef createTupleVariableDef(BLangVariable tupleVar) {
 
         BLangTupleVariableDef varDefNode = (BLangTupleVariableDef) TreeBuilder.createTupleVariableDefinitionNode();
-        varDefNode.pos = getPosition(null);
+        varDefNode.pos = tupleVar.pos;
         varDefNode.setVariable(tupleVar);
         return varDefNode;
     }
@@ -2815,7 +2816,7 @@ public class BLangNodeTransformer extends NodeTransformer<BLangNode> {
     private BLangErrorVariableDef createErrorVariableDef(BLangVariable errorVar) {
 
         BLangErrorVariableDef varDefNode = (BLangErrorVariableDef) TreeBuilder.createErrorVariableDefinitionNode();
-        varDefNode.pos = getPosition(null);
+        varDefNode.pos = errorVar.pos;
         varDefNode.setVariable(errorVar);
         return varDefNode;
     }
@@ -3411,19 +3412,19 @@ public class BLangNodeTransformer extends NodeTransformer<BLangNode> {
         List<BLangExpression> sizes = new ArrayList<>();
         Location position = getPosition(arrayTypeDescriptorNode);
         while (true) {
-            if (!arrayTypeDescriptorNode.arrayLength().isPresent()) {
-                sizes.add(new BLangLiteral(Integer.valueOf(OPEN_ARRAY_INDICATOR), symTable.intType));
+            if (arrayTypeDescriptorNode.arrayLength().isEmpty()) {
+                sizes.add(new BLangLiteral(OPEN_ARRAY_INDICATOR, symTable.intType));
             } else {
                 Node keyExpr = arrayTypeDescriptorNode.arrayLength().get();
                 if (keyExpr.kind() == SyntaxKind.NUMERIC_LITERAL) {
-                    BasicLiteralNode numericLiteralNode = (BasicLiteralNode) keyExpr;
-                    if (numericLiteralNode.literalToken().kind() == SyntaxKind.DECIMAL_INTEGER_LITERAL_TOKEN) {
-                        sizes.add(new BLangLiteral(Integer.parseInt(keyExpr.toString()), symTable.intType));
+                    Token literalToken = ((BasicLiteralNode) keyExpr).literalToken();
+                    if (literalToken.kind() == SyntaxKind.DECIMAL_INTEGER_LITERAL_TOKEN) {
+                        sizes.add(new BLangLiteral(Integer.parseInt(literalToken.text()), symTable.intType));
                     } else {
-                        sizes.add(new BLangLiteral(Integer.parseInt(keyExpr.toString(), 16), symTable.intType));
+                        sizes.add(new BLangLiteral(Integer.parseInt(literalToken.text(), 16), symTable.intType));
                     }
                 } else if (keyExpr.kind() == SyntaxKind.ASTERISK_LITERAL) {
-                    sizes.add(new BLangLiteral(Integer.valueOf(INFERRED_ARRAY_INDICATOR), symTable.intType));
+                    sizes.add(new BLangLiteral(INFERRED_ARRAY_INDICATOR, symTable.intType));
                 } else {
                     sizes.add(createExpression(keyExpr));
                 }

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/ParserTestRunner.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/ParserTestRunner.java
@@ -49,9 +49,6 @@ public class ParserTestRunner {
         HashSet<String> hashSet = new HashSet<>();
         // Following should be fixed
         // TODO: categorize and create issues
-        hashSet.add("array_type_source_17.bal");
-        hashSet.add("singleton_type_source_08.bal");
-        hashSet.add("singleton_type_source_02.bal");
         hashSet.add("xml_template_source_24.bal");
         hashSet.add("explicit-new-with-object-keyword-with-multiple-args-negative02.bal");
         hashSet.add("explicit-new-with-object-keyword-with-one-arg-negative01.bal");
@@ -69,7 +66,6 @@ public class ParserTestRunner {
         hashSet.add("import_decl_source_13.bal");
         hashSet.add("ambiguity_source_29.bal");
         hashSet.add("typed_binding_patterns_source_16.bal");
-        hashSet.add("forEach_stmt_source_07.bal");
         hashSet.add("do_stmt_source_07.bal");
         return hashSet;
     }

